go/service/chat_boxer.go

```diff
@@ -67,44 +67,97 @@ func (b *chatBoxer) unboxMessageWithKey(msg chat1.MessageBoxed, key *keybase1.Cr
 		return keybase1.Message{}, errors.New("nil ServerHeader in MessageBoxed")
 	}
 
-	if err := b.verifyMessageBoxed(msg); err != nil {
+	// decrypt body
+	jsonBody, err := b.open(msg.BodyCiphertext, key)
+	if err != nil {
 		return keybase1.Message{}, err
 	}
-
-	unboxed := keybase1.Message{
-		ServerHeader: *msg.ServerHeader,
+	var body keybase1.BodyPlaintext
+	if err := json.Unmarshal(jsonBody, &body); err != nil {
+		return keybase1.Message{}, err
 	}
 
-	if len(msg.BodyCiphertext.N) != libkb.NaclDHNonceSize {
-		return keybase1.Message{}, libkb.DecryptBadNonceError{}
+	// decrypt header
+	jsonHeader, err := b.open(msg.HeaderCiphertext, key)
+	if err != nil {
+		return keybase1.Message{}, err
+	}
+	var header keybase1.HeaderPlaintext
+	if err := json.Unmarshal(jsonHeader, &header); err != nil {
+		return keybase1.Message{}, err
 	}
-	var nonce [libkb.NaclDHNonceSize]byte
-	copy(nonce[:], msg.BodyCiphertext.N)
 
-	plaintext, ok := secretbox.Open(nil, msg.BodyCiphertext.E, &nonce, ((*[32]byte)(&key.Key)))
-	if !ok {
-		return keybase1.Message{}, libkb.DecryptOpenError{}
+	// verify the message
+	if err := b.verifyMessage(header, msg); err != nil {
+		return keybase1.Message{}, err
 	}
 
-	if err := json.Unmarshal(plaintext, &unboxed.MessagePlaintext); err != nil {
+	// create a chat1.MessageClientHeader from versioned HeaderPlaintext
+	var clientHeader chat1.MessageClientHeader
+	headerVersion, err := header.Version()
+	if err != nil {
 		return keybase1.Message{}, err
 	}
+	switch headerVersion {
+	case keybase1.HeaderPlaintextVersion_V1:
+		hp := header.V1()
+		clientHeader = chat1.MessageClientHeader{
```

Comment:
this is kind of weird, to be converting from a versioned structure into an unversioned structure.  Is there a flaw in the protocol design at a higher level?